test/directives/prefetch.f90

```diff
@@ -11,13 +11,13 @@ subroutine prefetch_dir(a1, a2)
   end do
 end subroutine prefetch_dir
 
-!! Ensure that the offset generated for the prefetch of a2(i + 256) is correct.
 ! CHECK-PREFETCH: [[a1:%[0-9]+]] = bitcast i64* %a1 to i8*
 ! CHECK-PREFETCH: [[a2:%[0-9]+]] = bitcast i64* %a2 to i8*
-! CHECK-PREFETCH: [[a2base:%[0-9]+]] = getelementptr i8, i8* [[a2]], i64 1020
 ! CHECK-PREFETCH: call void @llvm.prefetch{{.*}}(i8* [[a1]], i32 0, i32 3, i32 1)
 ! CHECK-PREFETCH: [[i:%[0-9]+]] = shl nuw nsw i64 %indvars.iv, 2
-! CHECK-PREFETCH: [[a2elem:%[0-9]+]] = getelementptr i8, i8* [[a2base]], i64 [[i]]
+!! Ensure that the offset generated for the prefetch of a2(i + 256) is correct.
+! CHECK-PREFETCH: [[a2offidx:%.*]] = add nuw nsw i64 [[i]], 1020
```

Comment:
Is there another way to write this expression? I'm getting here an error with release_12x:

flang/test/directives/prefetch.f90:19:19: error: CHECK-PREFETCH: expected string not found in input
! CHECK-PREFETCH: [[a2offidx:%.*]] = add nuw nsw i64 [[i]], 1020
                                     ^

Reply:
with the -O0 option I got this llvm (in both releases):
`; ModuleID = '/tmp/prefetch-a4d815.ll'
source_filename = "/tmp/prefetch-a4d815.ll"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

@.C305_prefetch_dir_ = internal constant i32 127
@.C304_prefetch_dir_ = internal constant i32 256
@.C285_prefetch_dir_ = internal constant i32 1
@.C303_prefetch_dir_ = internal constant i32 3968
@.C302_prefetch_dir_ = internal constant i32 128

define void @prefetch_dir_(i64* %a1, i64* %a2) !dbg !5 {
L.entry:
  %.dY0001_313 = alloca i32, align 4
  %i_301 = alloca i32, align 4
  br label %L.LB1_316

L.LB1_316:                                        ; preds = %L.entry
  store i32 3841, i32* %.dY0001_313, align 4, !dbg !12
  store i32 128, i32* %i_301, align 4, !dbg !12
  br label %L.LB1_311

L.LB1_311:                                        ; preds = %L.LB1_311, %L.LB1_316
  %0 = bitcast i64* %a1 to i8*, !dbg !13
  call void @llvm.prefetch.p0i8(i8* %0, i32 0, i32 3, i32 1), !dbg !13
  %1 = bitcast i64* %a2 to i8*, !dbg !13
  %2 = getelementptr i8, i8* %1, i64 1020, !dbg !13
  %3 = load i32, i32* %i_301, align 4, !dbg !13
  %4 = sext i32 %3 to i64, !dbg !13
  %5 = mul nsw i64 %4, 4, !dbg !13
  %6 = getelementptr i8, i8* %2, i64 %5, !dbg !13
  call void @llvm.prefetch.p0i8(i8* %6, i32 0, i32 3, i32 1), !dbg !13
  %7 = load i32, i32* %i_301, align 4, !dbg !14
  %8 = sext i32 %7 to i64, !dbg !14
  %9 = bitcast i64* %a2 to i8*, !dbg !14
  %10 = getelementptr i8, i8* %9, i64 504, !dbg !14
  %11 = bitcast i8* %10 to i32*, !dbg !14
  %12 = getelementptr i32, i32* %11, i64 %8, !dbg !14
  %13 = load i32, i32* %12, align 4, !dbg !14
  %14 = load i32, i32* %i_301, align 4, !dbg !14
  %15 = sext i32 %14 to i64, !dbg !14
  %16 = bitcast i64* %a2 to i8*, !dbg !14
  %17 = getelementptr i8, i8* %16, i64 -512, !dbg !14
  %18 = bitcast i8* %17 to i32*, !dbg !14
  %19 = getelementptr i32, i32* %18, i64 %15, !dbg !14
  %20 = load i32, i32* %19, align 4, !dbg !14
  %21 = add nsw i32 %13, %20, !dbg !14
  %22 = load i32, i32* %i_301, align 4, !dbg !14
  %23 = sext i32 %22 to i64, !dbg !14
  %24 = bitcast i64* %a1 to i8*, !dbg !14
  %25 = getelementptr i8, i8* %24, i64 -4, !dbg !14
  %26 = bitcast i8* %25 to i32*, !dbg !14
  %27 = getelementptr i32, i32* %26, i64 %23, !dbg !14
  store i32 %21, i32* %27, align 4, !dbg !14
  %28 = load i32, i32* %i_301, align 4, !dbg !15
  %29 = add nsw i32 %28, 1, !dbg !15
  store i32 %29, i32* %i_301, align 4, !dbg !15
  %30 = load i32, i32* %.dY0001_313, align 4, !dbg !15
  %31 = sub nsw i32 %30, 1, !dbg !15
  store i32 %31, i32* %.dY0001_313, align 4, !dbg !15
  %32 = load i32, i32* %.dY0001_313, align 4, !dbg !15
  %33 = icmp sgt i32 %32, 0, !dbg !15
  br i1 %33, label %L.LB1_311, label %L.LB1_344, !dbg !15

L.LB1_344:                                        ; preds = %L.LB1_311
  ret void, !dbg !16
}

; Function Attrs: inaccessiblemem_or_argmemonly nofree nosync nounwind willreturn
declare void @llvm.prefetch.p0i8(i8* nocapture readonly, i32 immarg, i32 immarg, i32) #0

attributes #0 = { inaccessiblemem_or_argmemonly nofree nosync nounwind willreturn }

!llvm.module.flags = !{!0, !1}
!llvm.dbg.cu = !{!2}

!0 = !{i32 2, !"Dwarf Version", i32 4}
!1 = !{i32 2, !"Debug Info Version", i32 3}
!2 = distinct !DICompileUnit(language: DW_LANG_Fortran90, file: !3, producer: " F90 Flang - 1.5 2017-05-01", isOptimized: false, flags: "'+flang -O0 -S -emit-llvm /home/ARM/flang/test/directives/prefetch.f90'", runtimeVersion: 0, emissionKind: FullDebug, enums: !4, retainedTypes: !4, globals: !4, imports: !4, nameTableKind: None)
!3 = !DIFile(filename: "/home/ARM/flang/test/directives/prefetch.f90", directory: "/home/ARM/flang/build")
!4 = !{}
!5 = distinct !DISubprogram(name: "prefetch_dir", scope: !2, file: !3, line: 4, type: !6, scopeLine: 4, spFlags: DISPFlagDefinition, unit: !2)
!6 = !DISubroutineType(types: !7)
!7 = !{null, !8, !8}
!8 = !DICompositeType(tag: DW_TAG_array_type, baseType: !9, size: 131072, align: 32, elements: !10)
!9 = !DIBasicType(name: "integer", size: 32, align: 32, encoding: DW_ATE_signed)
!10 = !{!11}
!11 = !DISubrange(lowerBound: 1, upperBound: 4096)
!12 = !DILocation(line: 8, column: 1, scope: !5)
!13 = !DILocation(line: 9, column: 1, scope: !5)
!14 = !DILocation(line: 10, column: 1, scope: !5)
!15 = !DILocation(line: 11, column: 1, scope: !5)
!16 = !DILocation(line: 12, column: 1, scope: !5)`

Reply:
This is the error:
```
/home/ARM/flang/test/directives/prefetch.f90:18:19: error: CHECK-PREFETCH: expected string not found in input
! CHECK-PREFETCH: call void @llvm.prefetch{{.*}}(i8* [[a1]], i32 0, i32 3, i32 1)
                  ^
<stdin>:27:41: note: scanning from here
 %2 = getelementptr i8, i8* %1, i64 1020, !dbg !13
                                        ^
<stdin>:27:41: note: with "a1" equal to "%0"
 %2 = getelementptr i8, i8* %1, i64 1020, !dbg !13
                                        ^
<stdin>:32:2: note: possible intended match here
 call void @llvm.prefetch.p0i8(i8* %6, i32 0, i32 3, i32 1), !dbg !13
 ^

Input file: <stdin>
Check file: /home/ARM/flang/test/directives/prefetch.f90

-dump-input=help explains the following input dump.

Input was:
<<<<<<
            .
            .
            .
           22:
           23: L.LB1_311: ; preds = %L.LB1_311, %L.LB1_316
           24:  %0 = bitcast i64* %a1 to i8*, !dbg !13
           25:  call void @llvm.prefetch.p0i8(i8* %0, i32 0, i32 3, i32 1), !dbg !13
           26:  %1 = bitcast i64* %a2 to i8*, !dbg !13
           27:  %2 = getelementptr i8, i8* %1, i64 1020, !dbg !13
check:18'0                                             X~~~~~~~~~~ error: no match found
check:18'1                                                         with "a1" equal to "%0"
           28:  %3 = load i32, i32* %i_301, align 4, !dbg !13
check:18'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           29:  %4 = sext i32 %3 to i64, !dbg !13
check:18'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           30:  %5 = mul nsw i64 %4, 4, !dbg !13
check:18'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           31:  %6 = getelementptr i8, i8* %2, i64 %5, !dbg !13
check:18'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           32:  call void @llvm.prefetch.p0i8(i8* %6, i32 0, i32 3, i32 1), !dbg !13
check:18'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:18'2      ?                                                                     possible intended match
           33:  %7 = load i32, i32* %i_301, align 4, !dbg !14
check:18'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           34:  %8 = sext i32 %7 to i64, !dbg !14
check:18'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           35:  %9 = bitcast i64* %a2 to i8*, !dbg !14
check:18'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           36:  %10 = getelementptr i8, i8* %9, i64 504, !dbg !14
check:18'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           37:  %11 = bitcast i8* %10 to i32*, !dbg !14
check:18'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            .
            .
            .
```

Reply:
Good that they emit the same output. You challenge now is to convert the CHECK-PREFETCH lines from the original test (on LLVM12) to be checking for the same key points in the new code as the old. This new set of CHECK lines ought to match for LLVM 13 too.

The key thing that needs checking are the two prefetches and the values they prefetch.

Reply:
The goal of this test, as stated in the source itself, is to "ensure that the offset generated for the prefetch of `a2(i + 256)` is correct." The IR you pasted above does generate the correct offset for the second `llvm.prefetch` intrinsic call. The FileCheck pattern mismatches are due to two reasons:

1. The first prefetch call is placed before the `bitcast` of `%a2` in your IR. The original FileCheck patterns expect it to appear after the `bitcast`. You can fix this by reordering the FileCheck patterns.
2. The offset calculation for the second prefetch call involves a `load` because now the test is compiled at `-O0`, and the induction variable is no longer kept in a register. The effect of the following two snippets of code should be the same:
    ```llvm
    ; original
    %5 = shl nuw nsw i64 %indvars.iv, 2
    ```
    ```llvm
    ; -O0
    %3 = load i32, i32* %i_301, align 4
    %4 = sext i32 %3 to i64
    %5 = mul nsw i64 %4, 4
    ```
    So you should be able to fix this by rewriting the FileCheck patterns to match the three lines below.

Reply:
I did the suggested changes and the test fails with classic_flang driver release_13x. Seems this particular test cannot serve both releases (12 and 13) without adjustments for each release. My suggestion would be to either let it fail on release_12x or remove the test.

Reply:
I thought you said that both LLVM 12 and LLVM 13 produced the same IR. How could the FileCheck patterns fail to match with one but not the other?

Reply:
My lack knowledge on FileCheck doesn't help here, I wonder if someone else could explain this